dm/syncer/syncer.go

```diff
@@ -2252,12 +2275,34 @@ func (s *Syncer) handleRowsEvent(ev *replication.RowsEvent, ec eventContext) err
 	startTime := time.Now()
 	for i := range dmls {
 		job := newDMLJob(jobType, sourceTable, targetTable, dmls[i], &ec)
-		err = s.addJobFunc(job)
+		err = s.handleJobFunc(job)
 		if err != nil {
 			return err
 		}
 	}
 	metrics.DispatchBinlogDurationHistogram.WithLabelValues(jobType.String(), s.cfg.Name, s.cfg.SourceID).Observe(time.Since(startTime).Seconds())
+
+	if len(sourceTable.Schema) != 0 {
+		// when in position-based replication, now events before table checkpoint is sent to queue. But in GTID-based
+		// replication events may share one GTID, so event whose end position is equal to table checkpoint may not be
+		// sent to queue. We may need event index in GTID to resolve it.
+		s.saveTablePoint(sourceTable, *ec.currentLocation)
+	}
+
+	failpoint.Inject("flushFirstJob", func() {
+		if waitJobsDone {
+			s.tctx.L().Info("trigger flushFirstJob")
+			waitJobsDone = false
+
+			err2 := s.flushJobs()
+			if err2 != nil {
+				s.tctx.L().DPanic("flush checkpoint failed", zap.Error(err2))
+			}
+			failpoint.Return(nil)
+		}
+	})
+
+	s.checkShouldFlush()
```

Comment:
It seems that we run `s.checkShouldFlush()` twice for dmls?

Reply:
fixed in https://github.com/pingcap/tiflow/pull/3891/commits/e549451fae295da5a31ddcbd923d27bd74c19933